build.cake

```diff
@@ -116,20 +118,37 @@ Task("Publish")
     .WithCriteria(BuildSystem.IsRunningOnTeamCity)
     .Does(() =>
 {
-    NuGetPush($"{artifactsDir}Octostache.{nugetVersion}.nupkg", new NuGetPushSettings {
-        Source = "https://f.feedz.io/octopus-deploy/dependencies/nuget",
-        ApiKey = EnvironmentVariable("FeedzIoApiKey"),
-        SkipDuplicate = true,
+    var currentBranch = GitBranchCurrent(DirectoryPath.FromString(".")).FriendlyName;
+    var octopusServer = EnvironmentVariable("OctopusServerUrl");
+    var octopusApiKey = EnvironmentVariable("OctopusServerApiKey");
+    var space = EnvironmentVariable("OctopusServerSpaceName");
+    var octopusProjectName = "Octostache";
+
+    var nugetPackage = GetFiles($"{artifactsDir}Octostache.{nugetVersion}.nupkg");
+
+    // Current config for this repo doesn't generate prerelease tags, even if we're on a development/feature branch.
+    // Using the --overwrite-mode=IgnoreIfExists flag instructs the target Octopus server to ignore any attempted uploads with the same verison number.
+    // This decision is made server-side, so you'll still see log messages indicating the package was uploaded. Never fear, the push is discarded if the version already exists. 
+    // You can verify this by looking at the SHA1 and Published date of the package in the package feed: it won't change on subsequent pushes.
+    OctoPush(octopusServer, octopusApiKey, nugetPackage, new OctopusPushSettings 
+    {
+        ArgumentCustomization = args => args.Append("--overwrite-mode=IgnoreIfExists"),
+        Space = space 
     });
 
-    if (gitVersionInfo.PreReleaseTagWithDash == "")
-    {
-        NuGetPush($"{artifactsDir}Octostache.{nugetVersion}.nupkg", new NuGetPushSettings {
-            Source = "https://www.nuget.org/api/v2/package",
-            ApiKey = EnvironmentVariable("NuGetApiKey"),
-            SkipDuplicate = true,
-        });
-    }
+    // Config-as-Code doesn't yet support Automatic Release Creation, so do it manually
+    OctoCreateRelease(octopusProjectName, new CreateReleaseSettings {
+        ArgumentCustomization = args => args.Append($"--gitRef={currentBranch}"),
```

Comment:
Do we need to explicitly specify the git ref? Would it not build on the current branch that's checked out by default?

Reply:
This is about the Config-as-Code configuration. It's telling Octopus which branch's build configuration to use. Without this, it would default to master - at this time, there is no CaC definition for the project on master.

Additionally in the future if you want to test changes to the release pipeline on a branch, you would need this, otherwise it would just use the build plan from master as well (or at least, that's how I understand it - @emilol, would you be able to confirm that I've got this bit right? It's my first time CaC-ing :))

Reply:
I'll admit it's been a while since I've had my head in the CaC space, but I can have a crack at this one!

It looks like in this case you've got a repo being checked out on a build server and you're asking it to create an OD release based on the same repo. This instruction is telling the deploy server to checkout and use the same branch for deploying as you are for building.

It's not necessary for the CaC process to be committed to the same repo, nor is it assumed that you want to deploy from the same branch you're building from, that's up to the user which is why we ask for it to be specified.

We have an option to specify a default branch in the CaC deploy config, but I'm not sure if that's used for creating releases by default. I can vaguely recall there being reasons that that would be a dangerous assumption for our system to make, but I'm a bit fuzzy on the details there sorry.

@zentron how'd I do? 👋😄

Reply:
>you've got a repo being checked out on a build server and you're asking it to create an OD release based on the same repo. This instruction is telling the deploy server to checkout and use the same branch for deploying as you are for building

Spot on

Reply:
In fact @mik-ky, I think when I attempted to use the `CreateRelease` functionality, Octopus blocked me from doing so unless I specified that `--git-ref` parameter, because otherwise CaC doesn't know which build definition to use.

Reply:
Ah yep, makes sense - Thanks for the clarification @mjhilton @emilol !

Reply:
@emilol spot on! 
For additional context to one of @emilol points, We have decided not to default to the default branch for release creation to ensure users are explicitly aware of what they are deploying.

Reply:
Thankyou both. Good to know our understanding was correct.